### PR TITLE
feat(session): resolve self-registered sessions by surface UUID

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary

- Add the self-registration read side for `${CMUXLAYER_SESSION_REGISTRY:-$HOME/.cmuxlayer/session-registry.jsonl}` and wire it ahead of the deprecated transcript-directory scan in stdio, daemon, and app-server production entrypoints.
- Join records by the hook-provided `surface_uuid` (`CMUX_SURFACE_ID`) to `AgentRecord.surface_uuid`. Exact `launch_cwd` is only an optional secondary validator; newest epoch-ms wins remaining ties.
- Ignore `AgentRecord.pid` during selection. Production spawn, recovery, repair, and auto-discovery paths populate it as `null`; no production caller supplies the CLI PID through the generic transition seam.
- Require `session_id` + `surface_uuid`, tolerate missing/mismatched cwd and unknown fields, preserve `session_path`, reject non-integer PID/timestamp metadata, and fall back without throwing on malformed/missing registry data.

## Binding-key blocker resolved

The initial pid/cwd contract was unsound for RAW seats: live repaired/auto records had both `pid:null` and `launch_cwd:null`, while terminal metadata exposed the shell cwd rather than the agent worktree cwd. The design ruling is now `CMUX_SURFACE_ID -> AgentRecord.surface_uuid`, which is the same stable UUID space already used by registry reconciliation. Tests cover two records with absent/mismatched cwd and distinct UUIDs, proving each agent binds only its own session.

Records from an older writer that lack `surface_uuid` intentionally return null and use the bounded last-resort scan; layerSpec is adding `surface_uuid` to the writer record.

## Verification

- TDD: UUID-primary tests failed against cwd-first selection, then passed after the resolver/parser change.
- `bun run test tests/self-registration.test.ts` — 21/21 passed.
- `bun run typecheck && env -u CMUX_SOCKET_PATH -u CMUX_DAEMON_SOCKET bun run test` — 105/105 files, 2231/2231 tests passed on committed head.
- The push hook independently finished green (`run_tests.sh finished with exit status 0`; 2231/2231 Vitest tests plus contract/regression checks).

One declared spawn-manifest flake initially timed out because its static generated temp directory had accumulated 3,180 files / 148 MB. Moving that exact fixture directory to Trash made the unchanged focused test pass in 116 ms; two subsequent full gates were green. No timeout or coverage workaround was used.

Local CodeRabbit was attempted but its free OSS quota was exhausted; GitHub CI/review findings will be processed on this PR.

Base: `origin/main` / v0.4.15. Do not merge; held for the cleared Claude review and Etan's merge/release decision.

---

> [!NOTE]
> **Medium Risk**
> Changes core agent lifecycle session binding used during boot and sweeps; incorrect matching or fallback behavior could leave agents without `cli_session_id`, though transcript scan remains as last resort and tests default stays hermetic.
> 
> **Overview**
> Introduces **self-registration as the primary way** to bind managed agents to harness `session_id` values at boot, instead of inferring from `~/.claude` / `~/.codex` transcript scans.
> 
> A new **`self-registration.ts`** module reads append-only **`session-registry.jsonl`** (or `CMUXLAYER_SESSION_REGISTRY`), parses tolerant JSONL rows, and matches **`surface_uuid`** on the agent to hook-written records—optionally preferring exact **`launch_cwd`**, then newest **`ts`**. Production uses an **incremental index** (dev/inode, byte continuity, bounded per-surface history) so sweeps do not re-read the whole file.
> 
> **`AgentEngine`** gains **`selfRegistrationSessionResolver`** (off by default for hermetic tests). Resolution order is: self-registration → injected **`sessionIdentityResolver`** (test override) → deprecated **`findTranscriptSessionIdentity`**. Boot capture paths (`maybeCaptureBootSessionId`) honor the same ordering.
> 
> **Stdio, daemon, and app-server** entrypoints inject **`makeSelfRegistrationSessionResolver()`** through **`createServer`** / daemon context options; daemon tests assert custom resolver forwarding.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6397ba5e50327b541c39e0ac38a8e2b96a7bbffe. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Resolve self-registered sessions by surface UUID from a JSONL registry file
> - Adds `makeSelfRegistrationSessionResolver` in [self-registration.ts](https://github.com/EtanHey/cmuxlayer/pull/338/files#diff-cf32477d60bc1f3ca2d584c32ddbd325f3ec591ab42b5daf4d9b5ffebc76beea), which reads a JSONL registry file (default `~/.cmuxlayer/session-registry.jsonl`, overridable via `CMUXLAYER_SESSION_REGISTRY`) and returns a matching `session_id` and path keyed by `surface_uuid`.
> - The resolver uses an incremental, append-aware reader that detects file rotation/truncation and re-indexes only new bytes, avoiding full re-reads on each invocation.
> - `AgentEngine` now tries the self-registration resolver first, then an optional injected fallback, and only falls back to the legacy transcript scan as a last resort.
> - The resolver is wired into all three server entry points: `CmuxLayerDaemon`, `CmuxAppServerRuntime`, and `startInProcessRuntime`, with an escape hatch to inject a custom resolver via options.
> - Risk: the legacy transcript scan is now bypassed whenever self-registration succeeds, which changes session capture behavior for agents with a non-empty `surface_uuid` in a JSONL-backed harness.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 6397ba5.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Added self-registration–based session identity resolution as the primary matching path.
  * Exposed an optional custom session resolver across server and daemon setups, including in-process runtime.
  * Improved session registry handling for incremental updates, truncation, and safe candidate selection.

* **Bug Fixes**
  * Strengthened fallback behavior when self-registration can’t determine a session.
  * Refined boot prompt detection, readiness matching, and session capture/command status reporting.

* **Tests**
  * Added extensive unit and integration tests covering resolver behavior, registry edge cases, and engine integration.

### CODE REVIEW

File: src/self-registration.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+      text = readFile(registryPath);
+    } catch {
+      return null; // missing/unreadable → fall back to scan
+    }
+    if (!text) return null;
+
+    const candidates = parseSelfRegistrationLines(text).filter(
+      (entry) => surfaceUuidKey(entry.surface_uuid) === agentSurfaceUuid,
+    );
+    const chosen = chooseCandidate(candidates, launchCwd);
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts:513]: Fixed in 613524d. Same-surface candidates now require a safe epoch-ms timestamp at or after the agent creation window (with the existing five-second launch skew); stale and undated rows fail closed. Regression tests cover stale rejection, the exact tolerance boundary, and missing timestamps.

File: src/daemon.ts [hunk 2]
```diff
@@ -519,10 +520,11 @@ export class CmuxLayerDaemon {
           enableClaudeChannels: this.opts.enableClaudeChannels,
           spawnPreflight: this.opts.spawnPreflight,
           disableSpawnPreflight: this.opts.disableSpawnPreflight,
+          selfRegistrationSessionResolver:
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/daemon.ts:523]: Fixed in 613524d. Daemon lazy context now forwards the injected resolver and only constructs the HOME-backed default when no override was supplied. A daemon regression test asserts reference-identical forwarding.

File: src/agent-engine.ts [hunk 3]
```diff
@@ -1879,8 +1896,8 @@ export class AgentEngine {
     if (!JSONL_HARNESSES.has(agent.cli)) return false;
     const hasManagedLaunchContext = Boolean(
       agent.launcher_name ||
-        agent.launch_cwd?.trim() ||
-        agent.worktree_path?.trim(),
+      agent.launch_cwd?.trim() ||
+      agent.worktree_path?.trim(),
     );
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/agent-engine.ts:1917]: Fixed in 613524d. Self-registration eligibility is now independent of transcript-scan eligibility and requires only a capture state, JSONL harness, configured resolver, and stable surface UUID. The new raw auto-discovery regression has no launcher, cwd, or worktree and captures by UUID.

File: src/agent-engine.ts [hunk 4]
```diff
@@ -2021,6 +2036,33 @@ export class AgentEngine {
...
+   * bounded deferred-capture marker still guards that fallback.
+   */
+  private resolveSessionIdentityWithSelfRegistration(
+    agent: AgentRecord,
+  ): CapturedSessionIdentity | null {
+    const selfRegistered = this.selfRegistrationSessionResolver?.(agent);
+    if (selfRegistered) {
+      return this.normalizeCapturedSessionIdentity(selfRegistered);
+    }
+    return this.findTranscriptSessionIdentity(agent);
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/agent-engine.ts]: Fixed in 613524d. First-connect still suppresses transcript resolution, but directly invokes the eligible self-registration resolver. The regression test captures the registered session while asserting the transcript scan is never called.

**Comment by EtanHey** [comment]
@codex review

File: src/self-registration.ts [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    let text: string | null;
+    try {
+      text = readFile(registryPath);
+    } catch {
+      return null; // missing/unreadable → fall back to scan
+    }
+    if (!text) return null;
+
+    const candidates = parseSelfRegistrationLines(text).filter(
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in 3e53e16. The resolver now rejects entry timestamps beyond the reader wall clock plus the five-second skew, using an injectable clock. I intentionally bounded against resolve-time rather than createdAt: a legitimate hook can append after agent creation during a slow boot. RED/GREEN regressions prove a +5001ms row cannot dominate a current row and the exact +5000ms boundary remains accepted.

**Comment by EtanHey** [comment]
@codex review

File: src/self-registration.ts [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+      (entry) =>
+        surfaceUuidKey(entry.surface_uuid) === agentSurfaceUuid &&
+        entry.ts !== null &&
+        entry.ts >= earliestCurrentLaunchTs &&
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts:510]: Fixed in a05bdbd. The creation-time lower bound now applies only to surface_provenance=cmuxlayer_spawn, where created_at is the owned launch timestamp. Auto-discovered (unknown provenance) and resync-repaired (legacy/undefined provenance) rows skip that lower bound but retain UUID matching, timestamp presence, and the reader-clock upper bound. RED/GREEN tests cover both raw forms with a valid registration 60 seconds older than discovery; the managed stale-row regression still rejects beyond five seconds.

**Comment by EtanHey** [comment]
@codex review

File: src/self-registration.ts [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        ? createdAt - SESSION_REGISTRATION_TIMESTAMP_SKEW_MS
+        : Number.NEGATIVE_INFINITY;
+    const resolvedAt = now();
+    if (!Number.isSafeInteger(resolvedAt)) return null;
+    const latestPlausibleTs =
+      resolvedAt + SESSION_REGISTRATION_TIMESTAMP_SKEW_MS;
+
+    let text: string | null;
+    try {
+      text = readFile(registryPath);
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in 457a079. Production now keeps one append-aware, surface-UUID-indexed registry cache per resolver: unchanged sweeps perform only stat + indexed lookup, new bytes are range-read and parsed once, incomplete UTF-8/JSONL rows wait for their newline, and truncation/rotation/same-size rewrite reset cached candidates. Short reads fail closed without advancing the offset and retry on the next lookup. Regression coverage proves one initial read across multiple agents, delta-only appends, multibyte partial writes, same-size rewrite, inode rotation, truncation, and same-offset short-read retry. Exact full gate and push hook: 105/105 files, 2245/2245 tests.

File: src/self-registration.ts [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+    const rewrittenAtSameSize =
+      fileIdentity !== null &&
+      fileStat.size === consumedBytes &&
+      fileStat.mtimeMs !== observedMtimeMs;
+    if (
+      fileIdentity === null ||
+      replaced ||
+      fileStat.size < consumedBytes ||
+      rewrittenAtSameSize
+    ) {
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in cf65ddf with an append-continuity check that preserves the incremental-read guarantee. On same-inode growth, the reader now fetches the cached final 64 bytes plus the delta in one bounded range read and compares the overlap. A changed boundary triggers reset() plus a full byte-0 reindex, so removed candidates cannot survive; matching boundaries keep normal appends O(64-byte sentinel + delta). Short reads still fail closed without advancing state. The new regression rewrites the same inode to a larger body, proves new rows resolve, the stale row disappears, and recovery rereads from offset 0. Cleared Claude follow-up: GREEN. Full local gate and push hook: 105/105 files, 2246/2246 tests.

**Comment by EtanHey** [comment]
@codex review

File: src/self-registration.ts [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+    const resolvedAt = now();
+    if (!Number.isSafeInteger(resolvedAt)) return null;
+    const latestPlausibleTs =
+      resolvedAt + SESSION_REGISTRATION_TIMESTAMP_SKEW_MS;
+
+    const candidates = (readCandidates(agentSurfaceUuid) ?? []).filter(
+      (entry) =>
+        entry.ts !== null &&
+        entry.ts >= earliestCurrentLaunchTs &&
+        entry.ts <= latestPlausibleTs,
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts:511]: Fixed in 1eee4a4. Candidate selection now rejects a registry row whose explicit cli metadata (trimmed/lowercased) differs from agent.cli, so it fails closed to the existing fallback instead of binding a cross-harness stale session. Rows from older writers with no cli remain backward-compatible; surface_uuid stays primary, cwd stays optional secondary, and pid remains ignored. TDD: explicit Claude-row/Codex-agent mismatch went red then green; missing-cli compatibility is pinned. Cleared Claude narrow review: GREEN. Full local gate and push hook: 105/105 files, 2248/2248 tests.

**Comment by EtanHey** [comment]
@codex review

File: src/self-registration.ts [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+        }
+      }
+
+      consumedBytes = fileStat.size;
+      const continuitySource = Buffer.concat([continuityTail, appended]);
+      continuityTail = continuitySource.subarray(
+        Math.max(
+          0,
+          continuitySource.byteLength - SESSION_REGISTRATION_CONTINUITY_BYTES,
+        ),
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in 491962a. The cached continuity tail is now copied into an owned, bounded Buffer instead of retaining a subarray view; the post-newline pending fragment is copied as well. The regression asserts a distinct backing store and that mutating the source cannot change the cached tail. Local exact gate and the push hook both pass: typecheck + 105/105 files, 2249/2249 tests; cleared Claude narrow review: GREEN.

**Comment by EtanHey** [comment]
@codex review

Please review current head 491962a, especially the bounded owned-copy fix for the incremental registry reader continuity/pending buffers.

File: src/server.ts [hunk 11]
```diff
@@ -4163,124 +4165,129 @@ export function createServer(opts?: CreateServerOptions): McpServer {
...
-      observePtyWrite: true,
-      stableSurfaceIdentity: opts.stableSurfaceIdentity,
-    });
+      },
+      {
+        toolName: "send_command",
+        workspace: opts.workspace,
+        observePtyWrite: true,
+        stableSurfaceIdentity: opts.stableSurfaceIdentity,
+      },
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/server.ts:4289]: Verified against origin/main: this lock-key split predates PR #338 (the withSurfaceWrite identity logic was introduced before this branch; this PR only has inherited formatting churn at the cited area). It is a broader terminal-mutation concurrency design change spanning raw and identity-aware tools, unrelated to session self-registration, so expanding this PR would be unsafe. No change in this branch.

File: tests/self-registration.test.ts [hunk 12]
```diff
@@ -0,0 +1,935 @@
...
+        ): Promise<AgentRecord>;
+      }
+    ).maybeCaptureBootSessionId(rawAgent, {}, { resolveTranscript: false });
+
+    expect(captured.cli_session_id).toBe("sid-first-connect");
+    expect(selfRegistrationSessionResolver).toHaveBeenCalledTimes(1);
+    expect(scanSpy).not.toHaveBeenCalled();
+    engine.dispose();
+  });
+});
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/self-registration.test.ts:1286]: Keeping these cases in tests/self-registration.test.ts intentionally: they are feature-level acceptance tests for self-registration resolver composition and keep the new source module plus its AgentEngine wiring contract together. All cmux behavior is mocked and no live integration is used. Moving them into the already broad agent-engine suite would be organizational churn without a correctness or coverage gain.

File: src/self-registration.ts [hunk 13]
```diff
@@ -0,0 +1,455 @@
...
+
+/** Copy a bounded tail without retaining the source Buffer's backing store. */
+export function copyBufferTail(
+  source: Buffer,
+  maxBytes: number,
+): Buffer<ArrayBuffer> {
+  const boundedMaxBytes =
+    Number.isSafeInteger(maxBytes) && maxBytes > 0 ? maxBytes : 0;
+  const length = Math.min(source.byteLength, boundedMaxBytes);
+  const copy = Buffer.allocUnsafe(length);
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in 8f65a23. copyBufferTail now uses Buffer.allocUnsafeSlow(length), and the regression requires tail.buffer.byteLength === tail.byteLength in addition to distinct backing/mutation independence. Exact local gate, cleared Claude rerun, and push hook all pass: 105/105 files, 2254/2254 tests.

File: src/agent-engine.ts [hunk 14]
```diff
@@ -280,6 +280,15 @@ export interface AgentEngineOptions {
...
   sessionIdentityResolver?: SessionIdentityResolver;
+  /**
+   * PRIMARY session-identity resolver: the self-registration READ side. When
+   * provided (production entrypoints inject
+   * `makeSelfRegistrationSessionResolver()`), it is tried BEFORE the deprecated
+   * transcript scan and only falls through to the scan when it returns null.
+   * Default (bare/test construction): unset, so the engine touches no real
+   * registry file — hermetic like `outboxDrain`/`closeForensicsRunner`.
+   */
+  selfRegistrationSessionResolver?: SessionIdentityResolver;
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/agent-engine.ts:295]: Fixed the valid ordering issue in 8f65a23: sessionIdentityResolver is now composed after self-registration, so a self-registration hit wins and a miss invokes the injected fallback. The injected resolver intentionally remains an override for the real transcript scan (rather than enabling HOME I/O after returning null), preserving the existing hermetic test seam; when no override is supplied, the deprecated scan remains the final fallback. Tests cover all three paths. Full gate: 105/105 files, 2254/2254 tests; cleared Claude: GREEN.

File: src/self-registration.ts [hunk 15]
```diff
@@ -0,0 +1,455 @@
...
+        const completeLines = combined.subarray(0, finalNewline + 1);
+        pendingLine = Buffer.from(combined.subarray(finalNewline + 1));
+        for (const entry of parseSelfRegistrationLines(
+          completeLines.toString("utf8"),
+        )) {
+          const key = surfaceUuidKey(entry.surface_uuid);
+          if (!key) continue;
+          const existing = entriesBySurface.get(key);
+          if (existing) existing.push(entry);
+          else entriesBySurface.set(key, [entry]);
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in 8f65a23. Unterminated pending rows are retained only up to 64 KiB, and each surface keeps at most the 64 most recent appended registrations; malformed complete lines remain skipped and normal newest/cwd/CLI selection is unchanged over the recent set. Added regressions for both bounds. Full gate: 105/105 files, 2254/2254 tests; cleared Claude: GREEN.

**Comment by EtanHey** [comment]
@codex review

Final review requested on head 8f65a23 after addressing owned-buffer allocation, bounded pending/history state, and self-registration-first injected fallback composition.

File: src/self-registration.ts [hunk 16]
```diff
@@ -0,0 +1,480 @@
...
+): Buffer<ArrayBuffer> {
+  const boundedMaxBytes =
+    Number.isSafeInteger(maxBytes) && maxBytes > 0 ? maxBytes : 0;
+  const length = Math.min(source.byteLength, boundedMaxBytes);
+  const copy = Buffer.allocUnsafeSlow(length);
+  source.copy(copy, 0, source.byteLength - length);
+  return copy;
+}
+
+/** Retain a partial JSONL row only while it remains within the record cap. */
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts:129]: Fixed in 9675edd. The incremental reader now enters an explicit discard-until-newline state when an unterminated row exceeds the 64 KiB cap, so a valid-looking suffix can never be parsed as a standalone row. The new regression proves the suffix is ignored and that parsing resumes with the first complete record after the delimiter. Exact-head local and push-hook gates: typecheck + 105/105 files, 2255/2255 tests; cleared Claude reviewer: GREEN.

**Comment by EtanHey** [comment]
@codex review\n\nPlease review the exact current head 9675edde4bf43c68fbda71be65b22b21e17558c3, including the bounded incremental JSONL reader, continuity/reset logic, discard-until-newline behavior for oversized rows, per-surface candidate cap, self-registration precedence, and surface_uuid-primary binding. Do not merge.

File: src/self-registration.ts [hunk 17]
```diff
@@ -0,0 +1,480 @@
...
+          combined.subarray(finalNewline + 1),
+        );
+        for (const entry of parseSelfRegistrationLines(
+          completeLines.toString("utf8"),
+        )) {
+          const key = surfaceUuidKey(entry.surface_uuid);
+          if (!key) continue;
+          const existing = entriesBySurface.get(key);
+          if (existing) appendBoundedSurfaceCandidate(existing, entry);
+          else entriesBySurface.set(key, [entry]);
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in 753bcdee2b8298389d21114db71727bd0ea2444b. The production index now caps distinct surface keys at 1,024 and refreshes Map insertion recency on every append, evicting the least-recent historical surface on overflow. Existing per-surface history remains capped at 64, so total resolver state is bounded in both dimensions. The regression fills the cap, refreshes surface-0, adds one new surface, and proves surface-0/newest bind while surface-1 is evicted. Exact-head local and push-hook gates: typecheck + 105/105 files, 2256/2256 tests; cleared Claude review: GREEN.

**Comment by EtanHey** [comment]
@codex review\n\nPlease review exact current head 753bcdee2b8298389d21114db71727bd0ea2444b. In particular verify the global 1,024-surface index cap, append-recency eviction, interaction with the per-surface cap, reset/rotation paths, and the strengthened overflow regression. Do not merge.

File: src/self-registration.ts [hunk 18]
```diff
@@ -0,0 +1,499 @@
...
+      const finalNewline = combined.lastIndexOf(0x0a);
+      if (finalNewline >= 0) {
+        const completeLines = combined.subarray(0, finalNewline + 1);
+        const trailingLine = combined.subarray(finalNewline + 1);
+        discardingOversizedLine =
+          trailingLine.byteLength >
+          SESSION_REGISTRATION_MAX_PENDING_LINE_BYTES;
+        pendingLine = boundPendingRegistrationLine(trailingLine);
+        for (const entry of parseSelfRegistrationLines(
+          completeLines.toString("utf8"),
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in 29b0b447dac9de77c60e0bd4785ad28f09930e7e. Production complete-line parsing now scans newline boundaries in Buffer bytes and rejects rows over 64 KiB before toString/JSON.parse/indexing; the injected text path enforces the same byte cap. The red-then-green regression proves a complete valid-looking oversized row cannot bind and the following valid row still does. Exact-head local and push-hook gates: typecheck + 105/105 files, 2257/2257 tests; cleared Claude review: GREEN.

File: src/self-registration.ts [hunk 19]
```diff
@@ -0,0 +1,499 @@
...
+        )) {
+          const key = surfaceUuidKey(entry.surface_uuid);
+          if (!key) continue;
+          const existing = entriesBySurface.get(key);
+          if (existing) appendBoundedSurfaceCandidate(existing, entry);
+          else entriesBySurface.set(key, [entry]);
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in 753bcdee2b8298389d21114db71727bd0ea2444b and retained in current head 29b0b447dac9de77c60e0bd4785ad28f09930e7e. Distinct surface keys are capped at 1,024 with least-recent append eviction; per-surface candidate arrays remain capped at 64. The overflow regression refreshes surface-0, adds a new key, and proves surface-0/newest survive while surface-1 evicts. Exact-head gates: 105/105 files, 2257/2257 tests.

**Comment by EtanHey** [comment]
@codex review\n\nPlease review exact current head 29b0b447dac9de77c60e0bd4785ad28f09930e7e. Focus on byte-bounded complete-row parsing, partial/discard state interaction, the global surface-key cap, and all prior finding fixes. Do not merge.

File: src/self-registration.ts [hunk 20]
```diff
@@ -0,0 +1,560 @@
...
+        const completeLines = combined.subarray(0, finalNewline + 1);
+        const trailingLine = combined.subarray(finalNewline + 1);
+        discardingOversizedLine =
+          trailingLine.byteLength >
+          SESSION_REGISTRATION_MAX_PENDING_LINE_BYTES;
+        pendingLine = boundPendingRegistrationLine(trailingLine);
+        for (const entry of parseSelfRegistrationBufferLines(completeLines)) {
+          const key = surfaceUuidKey(entry.surface_uuid);
+          if (!key) continue;
+          indexSurfaceCandidate(entriesBySurface, key, entry);
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in a1949f4f5766350a128ae50e68f25e8b95ec1a33. Rows with ts:null are now rejected before indexSurfaceCandidate, so universally non-resolvable rows cannot consume any of the 1,024 surface slots. Agent-dependent rows remain indexed and filtered at lookup time; I explicitly avoided a CLI prefilter after confirming JSONL_HARNESSES also includes Cursor. The red-then-green regression puts one valid older row before 1,024 distinct ts:null rows and proves the valid surface still binds. Exact-head local and push-hook gates: typecheck + 105/105 files, 2258/2258 tests; cleared Claude review: GREEN.

**Comment by EtanHey** [comment]
@codex review\n\nPlease review exact current head a1949f4f5766350a128ae50e68f25e8b95ec1a33. Verify all prior finding fixes, especially pre-cap rejection of ts:null rows without dropping agent-dependent candidates, bounded surface/history state, and oversized-row handling. Do not merge.

File: src/self-registration.ts [hunk 21]
```diff
@@ -0,0 +1,566 @@
...
+      reset();
+    }
+    fileIdentity = nextIdentity;
+
+    if (fileStat.size > consumedBytes) {
+      const continuityLength = sameFileGrowth
+        ? Math.min(continuityTail.byteLength, consumedBytes)
+        : 0;
+      const offset = consumedBytes - continuityLength;
+      const length = fileStat.size - offset;
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/self-registration.ts]: Fixed in 6397ba5e50327b541c39e0ac38a8e2b96a7bbffe. Cold-start, reset, and append reads now stream through fixed 64 KiB chunks; continuity remains a separate 64-byte read. The reader advances only after each complete chunk, so a short/error read preserves the already indexed prefix and safely resumes after continuity verification. Added a regression with a >3-chunk oversized row, an injected short second chunk, and a later valid record; it proves every requested range is bounded and the next resolve resumes successfully. Local full gate and the intact push hook both pass 105 files / 2259 tests; cleared Claude inner-loop review is GREEN.

**Comment by EtanHey** [comment]
@codex review exact head 6397ba5e50327b541c39e0ac38a8e2b96a7bbffe. The previous cold-start full-file-read finding is fixed with resumable 64 KiB streaming reads and a large-registry/short-read regression; please review the exact pushed head.

**Comment by EtanHey** [comment]
## SETTLED — final head `6397ba5e50327b541c39e0ac38a8e2b96a7bbffe`

Base: `main` at annotated tag `v0.4.15` (`c903cca`). PR is MERGEABLE/CLEAN. **Do not merge.**

### Binding / PID finding

- Production spawn, recovery, repair, and auto-registration paths populate `AgentRecord.pid` as `null`; the generic transition seam can accept a PID, but no production caller supplies a reliable CLI-process PID. Numeric PIDs are confined to tests/control-health evidence. Therefore PID is unsound for identity selection and is ignored.
- Primary binding is exact normalized `entry.surface_uuid === agent.surface_uuid` (the `CMUX_SURFACE_ID` UUID space).
- Exact launch cwd is optional secondary evidence within a matching surface; newest valid epoch-ms record wins, with append order breaking equal-timestamp ties.

### Read-side robustness

- Self-registration resolves before injected legacy identity and transcript-scan fallback.
- Production JSONL reads are incremental and resumable: cold start, reset, and append paths use at most 64 KiB per data read, with a separate <=64-byte continuity check.
- Short/error reads retain the already indexed prefix and resume safely; continuity mismatch resets and re-streams from byte 0.
- State is bounded at 64 KiB pending row, 64 candidates/surface, and 1,024 indexed surfaces; complete and partial oversized rows are discarded safely, and `ts:null` rows are rejected before consuming the global cap.
- Regression coverage includes two mismatched/null-cwd records with distinct surface UUIDs binding independently, large cold registries, oversized rows, short-read resume, rotation/rewrite continuity, caps, freshness, CLI compatibility, and fallback precedence.

### Final gate / review receipt

- Local required gate: `bun run typecheck && env -u CMUX_SOCKET_PATH -u CMUX_DAEMON_SOCKET bun run test` — GREEN, 105 files / 2259 tests. One known `fleet-sidebar` timing flake was reproduced green focused and the complete gate then passed.
- Intact push hook — GREEN, 105 files / 2259 tests after the same known `fleet-sidebar` flake rerun; never bypassed.
- GitHub CI — GREEN on exact head. First attempt hit the unrelated `server.test.ts` Codex auto-update fake-timer load flake; exact focused test passed 20/20 locally, and the failed CI job rerun passed.
- build-site PASS; Macroscope PASS; CodeRabbit PASS. Cursor Bugbot skipped because its account usage limit was reached.
- GitHub Codex exact-head review: no major issues (`6397ba5e50`).
- Cleared Claude exact-head inner-loop: GREEN; traced bounded cold/reset/append reads, short-read resume, continuity mismatch recovery, allocation bounds, and duplicate prevention.
- Unresolved review threads: **0**.

Holding for Etan review/merge-release decision; this branch/PR has not been merged.
